routers/repo/blame.go

```diff
@@ -0,0 +1,231 @@
+		}
+	}
+
+	// Show latest commit info of repository in table header,
+	// or of directory if not in root directory.
+	latestCommit := ctx.Repo.Commit
+	if len(ctx.Repo.TreePath) > 0 {
+		latestCommit, err = ctx.Repo.Commit.GetCommitByPath(ctx.Repo.TreePath)
+		if err != nil {
+			ctx.ServerError("GetCommitByPath", err)
+			return
+		}
+	}
+	ctx.Data["LatestCommit"] = latestCommit
+	ctx.Data["LatestCommitVerification"] = models.ParseCommitWithSignature(latestCommit)
+	ctx.Data["LatestCommitUser"] = models.ValidateCommitWithEmail(latestCommit)
+
+	statuses, err := models.GetLatestCommitStatus(ctx.Repo.Repository, ctx.Repo.Commit.ID.String(), 0)
+	if err != nil {
+		log.Error(3, "GetLatestCommitStatus: %v", err)
+	}
+
+	// Get current entry user currently looking at.
+	entry, err := ctx.Repo.Commit.GetTreeEntryByPath(ctx.Repo.TreePath)
+	if err != nil {
+		ctx.NotFoundOrServerError("Repo.Commit.GetTreeEntryByPath", git.IsErrNotExist, err)
+		return
+	}
+
+	blob := entry.Blob()
+
+	ctx.Data["LatestCommitStatus"] = models.CalcCommitStatus(statuses)
+
+	ctx.Data["Paths"] = paths
+	ctx.Data["TreeLink"] = treeLink
+	ctx.Data["TreeNames"] = treeNames
+	ctx.Data["BranchLink"] = branchLink
+	ctx.Data["HighlightClass"] = highlight.FileNameToHighlightClass(entry.Name())
+	if !markup.IsReadmeFile(blob.Name()) {
+		ctx.Data["RequireHighlightJS"] = true
+	}
+	ctx.Data["RawFileLink"] = rawLink + "/" + ctx.Repo.TreePath
+	ctx.Data["PageIsViewCode"] = true
+
+	ctx.Data["IsBlame"] = true
+
+	if ctx.Repo.CanEnableEditor() {
+		ctx.Data["CanDeleteFile"] = true
+		ctx.Data["DeleteFileTooltip"] = ctx.Tr("repo.editor.delete_this_file")
+	} else if !ctx.Repo.IsViewBranch {
+		ctx.Data["DeleteFileTooltip"] = ctx.Tr("repo.editor.must_be_on_a_branch")
+	} else if !ctx.Repo.CanWrite(models.UnitTypeCode) {
+		ctx.Data["DeleteFileTooltip"] = ctx.Tr("repo.editor.must_have_write_access")
+	}
+
+	ctx.Data["FileSize"] = blob.Size()
+	ctx.Data["FileName"] = blob.Name()
+
+	blame, err := models.GetBlame(models.RepoPath(userName, repoName), commitID, fileName)
+
+	if err != nil {
+		ctx.NotFound("GetBlameCommit", err)
+		return
+	}
+
+	commitNames := make(map[string]string)
+
+	for _, part := range blame.Parts {
+		sha := part.Sha
+
+		commit, err := ctx.Repo.GitRepo.GetCommit(sha)
+		if err != nil {
+			if git.IsErrNotExist(err) {
+				ctx.NotFound("Repo.GitRepo.GetCommit", err)
+			} else {
+				ctx.ServerError("Repo.GitRepo.GetCommit", err)
+			}
+			return
+		}
+
+		commitNames[sha] = commit.CommitMessage
+
+	}
+
+	renderBlame(ctx, blame, commitNames)
+
+	ctx.HTML(200, tplBlame)
+}
+
+func renderBlame(ctx *context.Context, blame *models.BlameFile, commitNames map[string]string) {
+
+	repoLink := ctx.Repo.RepoLink
+
+	var lines = make([]string, 0, 0)
+
+	for _, part := range blame.Parts {
+		for _, line := range part.Lines {
+			lines = append(lines, line)
+		}
+	}
+
+	var output bytes.Buffer
+
+	var i = 0
+
+	for _, part := range blame.Parts {
```

Comment:
Surely the template can do this? You're iterating over the blame parts three times and rendering each bit into html in code. 

Use a `range` within the template. Instead of having lists within a table with one row, use `range` to create multiple rows with cells for each bit. Then you can avoid having to iterate over the file multiple times.

Reply:
There are a lot of `if last element:` that I don't think can be easily translated to template. I've moved everything to one for loop.